lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            ``False`` if it was pulled in as a dependency of an explicit
+            package.
+    """
+    if explicit and not rec.explicit:
+        with spack.store.db.write_transaction():
+            rec = spack.store.db.get_record(pkg.spec)
+            message = '{s.name}@{s.version} : marking the package explicit'
+            tty.msg(message.format(s=pkg.spec))
+            rec.explicit = True
+
+
+def dump_packages(spec, path):
+    """
+    Dump all package information for a spec and its dependencies.
+
+    This creates a package repository within path for every namespace in the
+    spec DAG, and fills the repos wtih package files and patch files for every
+    node in the DAG.
+
+    Args:
+        spec (Spec): the Spack spec whose package information is to be dumped
+        path (str): the path to the build packages directory
+    """
+    mkdirp(path)
+
+    # Copy in package.py files from any dependencies.
+    # Note that we copy them in as they are in the *install* directory
+    # NOT as they are in the repository, because we want a snapshot of
+    # how *this* particular build was done.
+    for node in spec.traverse(deptype=all):
+        if node is not spec:
+            # Locate the dependency package in the install tree and find
+            # its provenance information.
+            source = spack.store.layout.build_packages_path(node)
+            source_repo_root = os.path.join(source, node.namespace)
+
+            # There's no provenance installed for the source package.  Skip it.
+            # User can always get something current from the builtin repo.
+            if not os.path.isdir(source_repo_root):
+                continue
+
+            # Create a source repo and get the pkg directory out of it.
+            try:
+                source_repo = spack.repo.Repo(source_repo_root)
+                source_pkg_dir = source_repo.dirname_for_package_name(
+                    node.name)
+            except spack.repo.RepoError:
+                tty.warn("Warning: Couldn't copy in provenance for {0}"
+                         .format(node.name))
+
+        # Create a destination repository
+        dest_repo_root = os.path.join(path, node.namespace)
+        if not os.path.exists(dest_repo_root):
+            spack.repo.create_repo(dest_repo_root)
+        repo = spack.repo.Repo(dest_repo_root)
+
+        # Get the location of the package in the dest repo.
+        dest_pkg_dir = repo.dirname_for_package_name(node.name)
+        if node is not spec:
+            install_tree(source_pkg_dir, dest_pkg_dir)
+        else:
+            spack.repo.path.dump_provenance(node, dest_pkg_dir)
+
+
+def get_install_msg(name):
+    """
+    Colorize the name/id of the package being installed
+
+    Args:
+        name (str): Name/id of the package being installed
+
+    Return:
+        (str) Colorized installing message
+    """
+    return colorize('@*{Installing} @*g{%s}' % name)
```

Comment:
This function is only used 2 times. Would it be worth inlining it?

Reply:
True but stand-along functions are easier to test. :)